### PR TITLE
Added support for selecting files when using "Open containing folder"

### PR DESCRIPTION
**Pull request by DavidKarlas**
This is very useful when folder has many files since it highlights and scrolls to file(s) user is interested in...

Linux is messed up... There is DBus command for this: "org.freedesktop.FileManager1.ShowItems" but it seems only Nautilus supports it... I could probably special case(detect if nautilus is main file browser) or any other file browser(dolphin) but then... Nautilus doesn't support selecting multiple files(it opens multiple windows for files in same folder).

So for now Linux doesn't support this :(

### CODE REVIEW

File: main/src/addins/WindowsPlatform/WindowsPlatform/WindowsPlatform.cs [hunk 1]
```diff
@@ -399,6 +399,39 @@ static WindowsDesktopApplication WindowsAppFromName (string appName, bool isDefa
...
+		#region OpenFolder
+
+		[DllImport ("shell32.dll", ExactSpelling = true)]
+		static extern int SHOpenFolderAndSelectItems (
+			IntPtr pidlFolder,
+			uint cidl,
+			[In, MarshalAs (UnmanagedType.LPArray)] IntPtr[] apidl,
+			uint dwFlags);
+
+		[DllImport ("shell32.dll", CharSet = CharSet.Auto)]
```
Comment (Therzok) [inline main/src/addins/WindowsPlatform/WindowsPlatform/WindowsPlatform.cs]: No need to specify charset.

File: main/src/addins/WindowsPlatform/WindowsPlatform/WindowsPlatform.cs [hunk 2]
```diff
@@ -399,6 +399,39 @@ static WindowsDesktopApplication WindowsAppFromName (string appName, bool isDefa
...
+		static extern IntPtr ILCreateFromPath ([MarshalAs (UnmanagedType.LPTStr)] string pszPath);
+
+		[DllImport ("shell32.dll", CharSet = CharSet.None)]
+		static extern void ILFree (IntPtr pidl);
```
Comment (Therzok) [inline main/src/addins/WindowsPlatform/WindowsPlatform/WindowsPlatform.cs]: No need to specify charset.

File: main/src/addins/WindowsPlatform/WindowsPlatform/WindowsPlatform.cs [hunk 3]
```diff
@@ -399,6 +399,39 @@ static WindowsDesktopApplication WindowsAppFromName (string appName, bool isDefa
 			return null;
 		}
 
+		#region OpenFolder
+
+		[DllImport ("shell32.dll", ExactSpelling = true)]
```
Comment (Therzok) [inline main/src/addins/WindowsPlatform/WindowsPlatform/WindowsPlatform.cs]: This doesn't seem to be taking strings, so why do we want to force ExactSpelling = true?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/FileOperationsBuilderExtension.cs [hunk 4]
```diff
@@ -67,16 +67,33 @@ string GetDir (object ob)
...
-				FilePath path = GetDir (node.DataItem);
-
-				//if folder doesn't exist, walk up to parent that does
-				while (!path.IsNullOrEmpty && !path.IsDirectory)
-					path = path.ParentDirectory;
-
-				if (!path.IsNullOrEmpty && paths.Add (path))
-					DesktopService.OpenFolder (path);
+			var paths = new Dictionary<string, HashSet<string>> ();
+			foreach (var ob in CurrentNodes.Select((node)=>node.DataItem)) {
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/FileOperationsBuilderExtension.cs]: This line is specifically non-MD code-style. `CurrentNodes.Select (node => node.DataItem)` is better.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/FileOperationsBuilderExtension.cs [hunk 5]
```diff
@@ -67,16 +67,33 @@ string GetDir (object ob)
...
+				if (ob is IFileItem) {
+					string path = ((IFileItem)ob).FileName;
+					if (!string.IsNullOrEmpty (path)) {
+						var dirPath = (FilePath)System.IO.Path.GetDirectoryName (path);
+						//if folder doesn't exist, walk up to parent that does
+						while (!dirPath.IsNullOrEmpty && !dirPath.IsDirectory)
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/FileOperationsBuilderExtension.cs]: Hmmm, if I'm not mistaken, this is a duplicated piece of code in many places. Probably worth refactoring it into a extension method?

Comment (DavidKarlas) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Pads.ProjectPad/FileOperationsBuilderExtension.cs]: This should be addressed in different PR since changing it affects a lot of code unrelated to this PR

**Comment by DavidKarlas** [comment]
Ping, every time I use this on file which is in folder with 20+ files, I get hurt deep inside :'(

**Comment by Therzok** [comment]
Real men manually csproj files manually.
